common/blobstore/s3store/client.go

```diff
@@ -0,0 +1,311 @@
+		Key:    aws.String(key.String()),
+	})
+	if err != nil {
+		aerr, ok := err.(awserr.Error)
+		if ok && aerr.Code() == s3.ErrCodeNoSuchKey {
+			return false, nil
+		}
+		if ok && aerr.Code() == s3.ErrCodeNoSuchBucket {
+			return false, blobstore.ErrBucketNotExists
+		}
+
+		return false, err
+	}
+	return true, nil
+}
+
+func (c *client) Delete(ctx context.Context, bucket string, key blob.Key) (bool, error) {
+
+	_, err := c.s3cli.DeleteObjectWithContext(ctx, &s3.DeleteObjectInput{
+		Bucket: aws.String(bucket),
+		Key:    aws.String(key.String()),
+	})
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok {
+
+			if aerr.Code() == s3.ErrCodeNoSuchBucket {
+				return false, blobstore.ErrBucketNotExists
+			}
+
+			if aerr.Code() == s3.ErrCodeNoSuchKey {
+				return false, nil
+			}
+
+		}
+		return false, err
+
+	}
+	return true, nil
+}
+
+func (c *client) ListByPrefix(ctx context.Context, bucket string, prefix string) ([]blob.Key, error) {
+
+	results, err := c.s3cli.ListObjectsV2WithContext(ctx, &s3.ListObjectsV2Input{
+		Bucket: aws.String(bucket),
+		Prefix: aws.String(prefix),
+	})
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok && aerr.Code() == s3.ErrCodeNoSuchBucket {
+			return nil, blobstore.ErrBucketNotExists
+		}
+		return nil, ErrListFiles
+	}
+	var keys = make([]blob.Key, len(results.Contents))
+	for i, v := range results.Contents {
+		key, err := blob.NewKeyFromString(*v.Key)
+		if err != nil {
+			return nil, ErrConstructKey
+		}
+		keys[i] = key
+	}
+	return keys, nil
+}
+
+func (c *client) BucketMetadata(ctx context.Context, bucket string) (*blobstore.BucketMetadataResponse, error) {
+
+	results, err := c.s3cli.GetBucketAclWithContext(ctx, &s3.GetBucketAclInput{
+		Bucket: aws.String(bucket),
+	})
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok && aerr.Code() == s3.ErrCodeNoSuchBucket {
+			return nil, blobstore.ErrBucketNotExists
+		}
+		return nil, err
+	}
+
+	lifecycleResults, err := c.s3cli.GetBucketLifecycleWithContext(ctx, &s3.GetBucketLifecycleInput{
+		Bucket: aws.String(bucket),
+	})
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok && aerr.Code() == s3.ErrCodeNoSuchBucket {
+			return nil, blobstore.ErrBucketNotExists
+		}
+		return nil, err
+	}
+	var retentionDays = math.MaxInt64
+	for _, v := range lifecycleResults.Rules {
+		if retentionDays > int(*v.Expiration.Days) {
+			retentionDays = int(*v.Expiration.Days)
+		}
+	}
+	return &blobstore.BucketMetadataResponse{
+		Owner:         *results.Owner.DisplayName,
+		RetentionDays: retentionDays,
+	}, nil
+}
+
+func (c *client) BucketExists(ctx context.Context, bucket string) (bool, error) {
+
+	_, err := c.s3cli.HeadBucketWithContext(ctx, &s3.HeadBucketInput{
+		Bucket: aws.String(bucket),
+	})
+
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok && aerr.Code() == s3.ErrCodeNoSuchBucket {
+			return false, nil
+		}
+		return false, err
+	}
+	return true, nil
+}
+
+func (c *client) IsRetryableError(err error) bool {
```

Comment:
In the case of filestore I return false for all error types because it just for testing locally. But when running in prod we will want S3 to have a correct policy around retryable vs non-retryable errors.

Reply:
The aws sdk has built-in retry logic, should we disable that one and move it out to this? https://github.com/aws/aws-sdk-go/blob/master/aws/client/default_retryer.go